asset/asset.go

```diff
@@ -943,6 +944,447 @@ type GroupKeyReveal interface {
+	// derived by combining the hash of a non-spendable leaf and the root
+	// hash of the custom tapscript subtree.
+	//
+	// If a custom tapscript subtree root hash is provided, we use it.
+	// Otherwise, we default to an empty hash (a zero-filled byte array).
+	emptyNonSpendLeaf, err := NewNonSpendableScriptLeaf(nil).Unpack()
+	if err != nil {
+		return lfn.Err[GroupKeyRevealTapscript](err)
+	}
+
+	// Compute the tweaked custom branch hash.
+	tweakedCustomBranchHash := TapBranchHash(
+		emptyNonSpendLeaf.TapHash(),
+		customRoot.UnwrapOr(chainhash.Hash{}),
+	)
+
+	// Next, we'll combine the tweaked custom branch hash with the genesis
+	// asset ID leaf hash to compute the final tapscript root hash.
+	//
+	// Construct a non-spendable tapscript leaf for the genesis asset ID.
+	assetIDLeaf, err := NewNonSpendableScriptLeaf(
+		genesisAssetID[:],
+	).Unpack()
+	if err != nil {
+		return lfn.Err[GroupKeyRevealTapscript](err)
+	}
+
+	// Compute final tapscript root hash. This is the root hash of the
+	// tapscript tree that is used to derive the asset group key.
+	rootHash := TapBranchHash(
+		assetIDLeaf.TapHash(), tweakedCustomBranchHash,
+	)
+
+	// Construct the custom subtree inclusion proof. This proof is required
+	// to spend custom tapscript leaves in the tapscript tree.
+	emptyNonSpendLeafHash := emptyNonSpendLeaf.TapHash()
+	assetIDLeafHash := assetIDLeaf.TapHash()
+
+	customSubtreeInclusionProof := bytes.Join(
+		[][]byte{
+			emptyNonSpendLeafHash[:],
+			assetIDLeafHash[:],
+		}, nil,
+	)
+
+	return lfn.Ok(GroupKeyRevealTapscript{
+		root:                        rootHash,
+		customSubtreeRoot:           customRoot,
+		customSubtreeInclusionProof: customSubtreeInclusionProof,
+	})
+}
+
+// Validate checks that the group key reveal tapscript is well-formed and
+// compliant.
+func (g *GroupKeyRevealTapscript) Validate(assetID ID) error {
+	// Compute the final tapscript root hash from the genesis asset ID and
+	// the custom tapscript subtree root hash.
+	tapscript, err := NewGroupKeyTapscriptRoot(
+		assetID, g.customSubtreeRoot,
+	).Unpack()
+	if err != nil {
+		return fmt.Errorf("failed to compute tapscript root hash: %w",
+			err)
+	}
+
+	// Ensure that the final tapscript root hash matches the computed root
+	// hash.
+	customRoot := g.customSubtreeRoot.UnwrapOr(chainhash.Hash{})
+
+	if !g.root.IsEqual(&tapscript.root) {
+		return fmt.Errorf("failed to derive tapscript root from "+
+			"internal key, genesis asset ID, and "+
+			"custom subtree root (expected_root=%s, "+
+			"computed_root=%s, custom_subtree_root=%s, "+
+			"genesis_asset_id=%x)",
+			g.root, tapscript.root, customRoot, assetID[:])
+	}
+
+	return nil
+}
+
+// GroupKeyRevealV1 is a version 1 group key reveal type for representing the
+// data used to derive and verify the tweaked key used to identify an asset
+// group.
+type GroupKeyRevealV1 struct {
+	// version is the version of the group key reveal.
```

Comment:
See my struct sketch above that combines the main struct definition and also TLV fields. Having to always do `.Val` can be seen as cumbersome, in which case we can expose getter methods.